sdk/core/azure-core-amqp/src/main/java/com/azure/core/amqp/implementation/ReceiversPumpingScheduler.java

```diff
@@ -0,0 +1,194 @@
+        // It requires adding the key to the azure-core known configuration properties, let's evaluate it separately.
+        final Optional<Integer> poolMaxSizeOverridden = Optional.ofNullable(System.getProperty("com.azure.core.amqp.receiversPumpingThreadPoolMaxSize"))
+            .map(m -> {
+                try {
+                    return Integer.parseInt(m);
+                } catch (NumberFormatException e) {
+                    // Use poolMaxSizeDefault (the initialization log below hints the size is chosen).
+                    return null;
+                }
+            });
+        final int poolMaxSize = poolMaxSizeOverridden.orElseGet(poolMaxSizeDefault);
+        this.inner = Schedulers.newBoundedElastic(poolMaxSize, TASK_QUEUE_CAP, NAME, IDLE_TTL_SECONDS, true);
+        (new ClientLogger(ReceiversPumpingScheduler.class)).atVerbose()
+            .log("Initialized common ReceiversPumpingScheduler(maxThreads={})", poolMaxSize);
+    }
+
+    /**
+     * Check any of the Reactor's common (cached) Scheduler to see if "VirtualTimeScheduler" (VTScheduler) is injected;
+     * if so, ReceiversPumpingScheduler won't create its backing Scheduler and cache it in INSTANCE.
+     * <p/>
+     * Problem: When running under 'StepVerifier.withVirtualTime', the Reactor-test infra reset the "Schedulers" factory
+     * such that VirtualTimeScheduler is returned from all Schedulers.new** APIs and from APIs to obtain common (cached)
+     * Scheduler instances. As part of the reset, the Reactor also internally clears its common (cached) Scheduler
+     * instances so that the next cache load loads VirtualTimeScheduler. Unfortunately, Reactor is not providing
+     * any hooks for such reset so that the application caching its own Scheduler can clear its cache (INSTANCE static
+     * member in ReceiversPumpingScheduler case) on reset. It leads to the problem of
+     * <ul>
+     *     <li>a test_1 using 'StepVerifier.create' populating the INSTANCE with a real Scheduler S and a following
+     *     test_2 using 'StepVerifier.withVirtualTime' end up using S rather than loading and using VirtualTimeScheduler.</li>
+     *     <li>a test_1 using 'StepVerifier.withVirtualTime' loads INSTANCE with VirtualTimeScheduler, dispose it
+     *     at the end of test_2 execution, and a following test_2 using 'StepVerifier.withVirtualTime' uses the old cached
+     *     disposed VirtualTimeScheduler.</li>
+     * </ul>
+     *
+     * @return {@code true} if the VirtualTimeScheduler is loaded, {@code false} otherwise.
+     */
+    private static boolean isVTSchedulerMode() {
+        // Is there a better way to check VTScheduler? / should we request the Reactor team for an API to detect it?
+        //
+        final Scheduler s = Schedulers.boundedElastic();
+        if (s instanceof Supplier<?> && ((Supplier<?>) s).get() instanceof Scheduler) {
+            return ((Supplier<?>) s).get().getClass()
```

Comment:
Wow all those casts. Lol. :)

Reply:
:) It would have much simpler with Java-14 instance of pattern-matching [jep](https://openjdk.org/jeps/305)

```java
if (Schedulers.boundedElastic() instanceof Supplier<?> b && b.get() instanceof Scheduler s) {
    return s.getClass().getSimpleName().equals("VirtualTimeScheduler");
 }
```